controller/hybridgateway/managedfields/extract.go

```diff
@@ -0,0 +1,82 @@
+package managedfields
+
+import (
+	"bytes"
+	"fmt"
+
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v4/typed"
+)
+
+// ExtractAsUnstructured extracts the managed fields for a given field manager and subresource from a runtime.Object,
+// returning an unstructured.Unstructured containing only the fields managed by that manager.
+// Returns nil if no managed fields entry is found, or an error if extraction fails.
+func ExtractAsUnstructured(obj runtime.Object, fieldManager string, subresource string) (*unstructured.Unstructured, error) {
+	objectType, err := GetObjectType(obj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get object type for managed fields extraction: %w", err)
+	}
+	typedObj, err := toTyped(obj, objectType)
+	if err != nil {
+		return nil, fmt.Errorf("error converting obj to typed: %w", err)
+	}
+
+	accessor, err := apimeta.Accessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("error accessing metadata: %w", err)
+	}
+	fieldsEntry, ok := findManagedFields(accessor, fieldManager, subresource)
+	if !ok {
+		return nil, nil
+	}
+	fieldset := &fieldpath.Set{}
+	err = fieldset.FromJSON(bytes.NewReader(fieldsEntry.FieldsV1.Raw))
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling FieldsV1 to JSON: %w", err)
+	}
+
+	u := typedObj.ExtractItems(fieldset.Leaves()).AsValue().Unstructured()
+	m, ok := u.(map[string]any)
+	if !ok {
+		return nil, fmt.Errorf("unable to convert managed fields for %s to unstructured, expected map, got %T", fieldManager, u)
+	}
+
+	// TODO(alacuku): Properly set the type meta on the returned object inferring from the input object and scheme.
+	// set the type meta manually if it doesn't exist to avoid missing kind errors
+	// when decoding from unstructured JSON
+	if _, ok := m["kind"]; !ok && obj.GetObjectKind().GroupVersionKind().Kind != "" {
+		m["kind"] = obj.GetObjectKind().GroupVersionKind().Kind
+		m["apiVersion"] = obj.GetObjectKind().GroupVersionKind().GroupVersion().String()
+	}
```

Comment:
Can you elaborate on the scope of this TODO? Will it introduce any known issues in the logic of the current PR, or is it just a reminder for future extensibility?

Reply:
Removed the TOD and simplified the logic. It is just a matter of setting the gvk to the new unstructured resources from the original object. We are sure that it has the gvk set otherwise we do not get at this point.